File: tests/visualization/2DWP90_msgwam.py
import subprocess
import numpy as np
import matplotlib.pyplot as plt
import tools
import style

# Get host and user name.
host_name = subprocess.getoutput("hostname")
user_name = subprocess.getoutput("whoami")

if "levante" in host_name:
  # Levante cluster
  data_path = "/scratch/b/" + user_name + "/PF/runs"
  reference_path = "/scratch/b/" + user_name + "/PF/pinc/reference"

elif "login" in host_name:
  # Goethe cluster
  data_path = "/scratch/atmodynamics/" + user_name + "/PF/runs"
  reference_path = "/scratch/atmodynamics/" + user_name + "/PF/pinc/reference"

else:
  # Local machine
  data_path = "../"
  reference_path = "../"

print("data_path =", data_path)
print("reference_path =", reference_path)

# Import data.
data = tools.ModelOutput(data_path + '2DWP90_msgwam/')
reference = tools.ModelOutput(reference_path + '2DWP90_msgwam/')

# Set slices.
it = - 1
iz = slice(15, 45)

print(" ".join(("Time:", str(data.variables['t'][it]), "s")))

# Set grid.
xx = data.variables['x'][:] * 0.001
zz = data.variables['z'][iz] * 0.001

# Get large-scale tracer distribution change (tLS)
# and zonal, meridional, and vertical wind (uLS, vLS, wLS)
tLS = data.groups['atmvar'].variables['tmrd'][it, iz, 0]
uLS = data.groups['atmvar'].variables['u'][it, iz, 0]
vLS = data.groups['atmvar'].variables['v'][it, iz, 0]
wLS = data.groups['atmvar'].variables['w'][it, iz, 0]

tLSref = reference.groups['atmvar'].variables['tmrd'][it, iz, 0]
uLSref = reference.groups['atmvar'].variables['u'][it, iz, 0]
vLSref = reference.groups['atmvar'].variables['v'][it, iz, 0]
wLSref = reference.groups['atmvar'].variables['w'][it, iz, 0]

deltat = tLS - tLSref
deltau = uLS - uLSref
deltav = vLS - vLSref
deltaw = wLS - wLSref

# Create plot.
(fig, axes) = plt.subplots(nrows = 2, ncols = 2, figsize = (8, 6))

for ax in axes.flatten():
  ax.set_xlabel(r'$x \left[\mathrm{km}\right]$')
  ax.set_ylabel(r'$z \left[\mathrm{km}\right]$')

axes[0, 0].set_title(r'Large-scale zonal wind $u \, \left[\mathrm{m \,' \
    r's^{- 1}}\right]$')
(levels, colormap) = style.symmetric_contours(uLS.min(), uLS.max())
pplot00 = axes[0, 0].contourf(xx, zz, uLS, levels, cmap = colormap)
fig.colorbar(pplot00, ax = axes[0, 0])

axes[0, 1].set_title(r'Large-scale meridional wind $v \, \left[\mathrm{m \,' \
    r's^{- 1}}\right]$')
(levels, colormap) = style.symmetric_contours(vLS.min(), vLS.max())
pplot01 = axes[0, 1].contourf(xx, zz, vLS, levels, cmap = colormap)
fig.colorbar(pplot01, ax = axes[0, 1])

axes[1, 0].set_title(r'Large-scale vertical wind $w \, \left[\mathrm{m \,' \
    r's^{- 1}}\right]$')
(levels, colormap) = style.symmetric_contours(wLS.min(), wLS.max())
pplot10 = axes[1, 0].contourf(xx, zz, wLS, levels, cmap = colormap)
fig.colorbar(pplot10, ax = axes[1, 0])

axes[1, 1].set_title('Large-scale tracer distribution')
(levels, colormap) = style.symmetric_contours(tLS.min(), tLS.max())
pplot11 = axes[1, 1].contourf(xx, zz, tLS, levels, cmap = colormap)
fig.colorbar(pplot11, ax = axes[1, 1])

fig.savefig("".join((data_path, "/results/2DWP90_msgwam.pdf")))
fig.savefig("".join((data_path, "/results/2DWP90_msgwam.png")), dpi = 500)

# Create difference plot.
if data_path != reference_path:
  fig, axes = plt.subplots(nrows = 2, ncols = 2, figsize = (8, 6))

  for ax in axes.flatten():
    ax.set_xlabel(r'$x \left[\mathrm{km}\right]$')
    ax.set_ylabel(r'$z \left[\mathrm{km}\right]$')

  axes[0, 0].set_title(r'Large-scale zonal wind $u \, \left[\mathrm{m \,' \
      r's^{- 1}}\right]$')
  (levels, colormap) = style.symmetric_contours((deltau).min(), (deltau).max())
  pplot00 = axes[0, 0].contourf(xx, zz, deltau, levels, cmap = colormap)
  fig.colorbar(pplot00, ax = axes[0, 0])

  axes[0, 1].set_title(r'Large-scale meridional wind $v \, \left[\mathrm{m \,' \
      r's^{- 1}}\right]$')
  (levels, colormap) = style.symmetric_contours((deltav).min(), (deltav).max())
  pplot01 = axes[0, 1].contourf(xx, zz, deltav, levels, cmap = colormap)
  fig.colorbar(pplot01, ax = axes[0, 1])

  axes[1, 0].set_title(r'Large-scale vertical wind $w \, \left[\mathrm{m \,' \
      r's^{- 1}}\right]$')
  (levels, colormap) = style.symmetric_contours((deltaw).min(), (deltaw).max())
  pplot10 = axes[1, 0].contourf(xx, zz, deltaw, levels, cmap = colormap)
  fig.colorbar(pplot10, ax = axes[1, 0])

  axes[1, 1].set_title('Large-scale tracer distribution')
  (levels, colormap) = style.symmetric_contours((deltat).min(), (deltat).max())
  pplot11 = axes[1, 1].contourf(xx, zz, deltat, levels, cmap = colormap)
  fig.colorbar(pplot11, ax = axes[1, 1])

  fig.savefig("".join((data_path, "/results/2DWP90_msgwam_difference.pdf")))
  fig.savefig("".join((data_path, "/results/2DWP90_msgwam_difference.png")), \
      dpi = 500)

File: tests/visualization/2DWP90_msgwam.jl
include("style.jl")
include("tools.jl")

using LaTeXStrings

# Set paths.
data_path = "../"
reference_path = "../"

println("data_path =", data_path)
println("reference_path =", reference_path)

# Import data.
data = ModelOutput(data_path * "2DWP90_msgwam")
reference = ModelOutput(reference_path * "2DWP90_msgwam")

println("Time: ", data.variables["t"][end], " s")

# Set grid.
x = data.variables["x"][:] .* 0.001
z = data.variables["z"][15:45] .* 0.001
x = [xi for xi in x, j in 1:size(z)[1]]
z = [zj for i in 1:size(x)[1], zj in z]

# Get large-scale tracer distribution change (tls)
# and zonal, meridional, and vertical wind (uls, vls, wls)
tls = data.groups["atmvar"].variables["tmrd"][:, 1, 15:45, end]
uls = data.groups["atmvar"].variables["u"][:, 1, 15:45, end]
vls = data.groups["atmvar"].variables["v"][:, 1, 15:45, end]
wls = data.groups["atmvar"].variables["w"][:, 1, 15:45, end]

# Get reference data.
tlsr = reference.groups["atmvar"].variables["tmrd"][:, 1, 15:45, end]
ulsr = reference.groups["atmvar"].variables["u"][:, 1, 15:45, end]
vlsr = reference.groups["atmvar"].variables["v"][:, 1, 15:45, end]
wlsr = reference.groups["atmvar"].variables["w"][:, 1, 15:45, end]

# Compute differences.
deltat = tls .- tlsr
deltau = uls .- ulsr
deltav = vls .- vlsr
deltaw = wls .- wlsr

# Create figure.
figure(; figsize = (8, 6))

# Plot the zonal wind.
(levels, colormap) = symmetric_contours(minimum(uls), maximum(uls))
subplot(221)
zonal_wind = contourf(x, z, uls; levels = levels, cmap = colormap)
xlabel(L"x\,\left[\mathrm{km}\right]")
ylabel(L"z\,\left[\mathrm{km}\right]")
title(L"Large-scale zonal wind $u\,\left[\mathrm{m\,s^{-1}}\right]$")
colorbar(zonal_wind)

# Plot the meridional wind.
(levels, colormap) = symmetric_contours(minimum(vls), maximum(vls))
subplot(222)
meridional_wind = contourf(x, z, vls; levels = levels, cmap = colormap)
xlabel(L"x\,\left[\mathrm{km}\right]")
ylabel(L"z\,\left[\mathrm{km}\right]")
title(L"Large-scale meridional wind $v\,\left[\mathrm{m\,s^{-1}}\right]$")
colorbar(meridional_wind)

# Plot the vertical wind.
(levels, colormap) = symmetric_contours(minimum(wls), maximum(wls))
subplot(223)
vertical_wind = contourf(x, z, wls; levels = levels, cmap = colormap)
xlabel(L"x\,\left[\mathrm{km}\right]")
ylabel(L"z\,\left[\mathrm{km}\right]")
title(L"Large-scale vertical wind $w\,\left[\mathrm{m\,s^{-1}}\right]$")
colorbar(vertical_wind)

# Plot the tracer.
(levels, colormap) = symmetric_contours(minimum(tls), maximum(tls))
subplot(224)
tracer = contourf(x, z, tls; levels = levels, cmap = colormap)
xlabel(L"x\,\left[\mathrm{km}\right]")
ylabel(L"z\,\left[\mathrm{km}\right]")
title("Large-scale tracer")
colorbar(tracer)

# Save the figure.
savefig(data_path * "results/2DWP90_msgwam.png")

# Create difference plots.
if data_path != reference_path
  clf()

  # Plot the zonal-wind-differences.
  (levels, colormap) = symmetric_contours(minimum(deltauls), maximum(deltauls))
  subplot(221)
  zonal_wind = contourf(x, z, deltauls; levels = levels, cmap = colormap)
  xlabel(L"x\,\left[\mathrm{km}\right]")
  ylabel(L"z\,\left[\mathrm{km}\right]")
  title(L"Large-scale zonal wind $u\,\left[\mathrm{m\,s^{-1}}\right]$")
  colorbar(zonal_wind)

  # Plot the meridional-wind differences.
  (levels, colormap) = symmetric_contours(minimum(deltavls), maximum(deltavls))
  subplot(222)
  meridional_wind = contourf(x, z, deltavls; levels = levels, cmap = colormap)
  xlabel(L"x\,\left[\mathrm{km}\right]")
  ylabel(L"z\,\left[\mathrm{km}\right]")
  title(L"Large-scale meridional wind $v\,\left[\mathrm{m\,s^{-1}}\right]$")
  colorbar(meridional_wind)

  # Plot the vertical-wind differences.
  (levels, colormap) = symmetric_contours(minimum(deltawls), maximum(deltawls))
  subplot(223)
  vertical_wind = contourf(x, z, deltawls; levels = levels, cmap = colormap)
  xlabel(L"x\,\left[\mathrm{km}\right]")
  ylabel(L"z\,\left[\mathrm{km}\right]")
  title(L"Large-scale vertical wind $w\,\left[\mathrm{m\,s^{-1}}\right]$")
  colorbar(vertical_wind)

  # Plot the tracer differences.
  (levels, colormap) = symmetric_contours(minimum(deltatls), maximum(deltatls))
  subplot(224)
  tracer = contourf(x, z, deltatls; levels = levels, cmap = colormap)
  xlabel(L"x\,\left[\mathrm{km}\right]")
  ylabel(L"z\,\left[\mathrm{km}\right]")
  title("Large-scale tracer")
  colorbar(tracer)

  # Save the figure.
  savefig(data_path * "results/2DWP90_msgwam_difference.png")
end
